File: Лаб6/task3/task3.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>

using namespace std;

void printMatrix(const vector<vector<double>>& m, const string& name) {
    cout << name << ":\n" << fixed << setprecision(4);
    for (auto& row : m) {
        cout << "| ";
        for (size_t j = 0; j < row.size(); j++) {
            if (j == row.size() - 1) cout << "| ";
            cout << setw(10) << row[j] << " ";
        }
        cout << "|\n";
    }
    cout << "\n";
}

vector<double> gauss(vector<vector<double>> A, vector<double> b) {
    int n = A.size();
    vector<vector<double>> aug(n, vector<double>(n + 1));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) aug[i][j] = A[i][j];
        aug[i][n] = b[i];
    }

    printMatrix(aug, "Исходная расширенная матрица");

    for (int i = 0; i < n; i++) {
        int mr = i;
        for (int k = i + 1; k < n; k++)
            if (abs(aug[k][i]) > abs(aug[mr][i])) mr = k;
        if (mr != i) swap(aug[i], aug[mr]);

        for (int k = i + 1; k < n; k++) {
            double c = aug[k][i] / aug[i][i];
            for (int j = i; j <= n; j++) aug[k][j] -= c * aug[i][j];
        }
    }

    for (int i = 0; i < n; i++) {
        double d = aug[i][i];
        for (int j = i; j <= n; j++) aug[i][j] /= d;
    }

    printMatrix(aug, "Матрица после прямого хода (единичная диагональ)");

    vector<double> x(n);
    cout << "Результаты обратного хода:\n";
    for (int i = n - 1; i >= 0; i--) {
        x[i] = aug[i][n];
        for (int j = i + 1; j < n; j++) x[i] -= aug[i][j] * x[j];
        cout << "x" << i + 1 << " = " << x[i] << "\n";
    }
    return x;
}

void checkDiagonalDominance(const vector<vector<double>>& A) {
    cout << "\nПроверка диагонального преобладания:\n";

    bool ok = true;

    for (int i = 0; i < A.size(); i++) {

        double diag = abs(A[i][i]);
        double sum = 0;

        for (int j = 0; j < A.size(); j++) {
            if (i != j)
                sum += abs(A[i][j]);
        }

        cout << "Строка " << i + 1 << ": |a[" << i + 1 << "][" << i + 1 << "]| = " << diag << ", сумма остальных = " << sum;

        if (diag > sum) cout << "  -> OK\n";
        else {
            cout << "  -> НЕ выполняется\n";
            ok = false;
        }
    }

    if (ok) cout << "\nУсловие сходимости выполняется.\n";
    else cout << "\nУсловие сходимости НЕ выполняется.\n";
}

vector<double> seidel(const vector<vector<double>>& A, const vector<double>& b, double eps = 1e-3, int maxIter = 100)
{
    int n = A.size();
    vector<double> x(n, 0.0);
    vector<double> old(n);
    cout << "-------------------------------------------------------------\n";
    cout << " N        x1        x2        x3        x4        eps\n";
    cout << "-------------------------------------------------------------\n";

    for (int iter = 1; iter <= maxIter; iter++) {
        old = x;
        for (int i = 0; i < n; i++) {
            double sum = b[i];

            for (int j = 0; j < n; j++) {
                if (j < i) sum -= A[i][j] * x[j];
                else if (j > i) sum -= A[i][j] * old[j]; 
            }
            x[i] = sum / A[i][i];
        }

        double err = 0.0;

        for (int i = 0; i < n; i++) err = max(err, abs(x[i] - old[i]));
        cout << setw(2) << iter << " " << setw(10) << fixed << setprecision(4) << x[0] << setw(10) << x[1] << setw(10) << x[2] << setw(10) << x[3] << setw(12) << err << "\n";

        if (err < eps) break;
    }
    return x;
}


int main() {
    setlocale(LC_ALL, ".UTF8");
    double M = 0.89, N = 0.12, P = -1.15;

    vector<vector<double>> A = {
        {M,     -0.04,  0.21,  -1.16},
        {0.25,  -1.23,  N,     -0.09},
        {-0.21, N,      0.8,   -0.13},
        {0.15,  -1.31,  0.06,  P}
    };
    vector<double> B = { -1.24, P, 2.56, M };

    cout << "\nМетод Гаусса:\n";
    vector<double> Xg = gauss(A, B);
    cout << "\nРешение (Гаусс): x1=" << Xg[0] << ", x2=" << Xg[1] << ", x3=" << Xg[2] << ", x4=" << Xg[3] << "\n\n";

    checkDiagonalDominance(A);

    cout << "\nМетод Зейделя:\n";
    vector<double> Xs = seidel(A, B);
    cout << "\nРешение (Зейдель): x1=" << Xs[0] << ", x2=" << Xs[1] << ", x3=" << Xs[2] << ", x4=" << Xs[3] << "\n";

    return 0;
}

File: Лаб6/task3/task3.rs
fn print_matrix(m: &[Vec<f64>], name: &str) {
    println!("{}:", name);
    for row in m {
        print!("|");
        for &val in row {
            print!("{:10.4}", val);
        }
        println!(" |");
    }
    println!();
}

fn gauss(a: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = a.len();
    let mut aug = vec![vec![0.0; n + 1]; n];
    
    for i in 0..n {
        for j in 0..n {
            aug[i][j] = a[i][j];
        }
        aug[i][n] = b[i];
    }

    print_matrix(&aug, "Исходная расширенная матрица");

    for i in 0..n {
        let mut mr = i;
        for k in i + 1..n {
            if aug[k][i].abs() > aug[mr][i].abs() {
                mr = k;
            }
        }
        if mr != i {
            aug.swap(i, mr);
        }

        for k in i + 1..n {
            let c = aug[k][i] / aug[i][i];
            for j in i..=n {
                aug[k][j] -= c * aug[i][j];
            }
        }
    }

    for i in 0..n {
        let d = aug[i][i];
        for j in i..=n {
            aug[i][j] /= d;
        }
    }

    print_matrix(&aug, "Матрица после прямого хода (единичная диагональ)");

    let mut x = vec![0.0; n];
    println!("Результаты обратного хода:");
    for i in (0..n).rev() {
        x[i] = aug[i][n];
        for j in i + 1..n {
            x[i] -= aug[i][j] * x[j];
        }
        println!("x{} = {:.4}", i + 1, x[i]);
    }
    x
}

fn check_diagonal_dominance(a: &[Vec<f64>]) {
    println!("\nПроверка диагонального преобладания:");
    let mut ok = true;
    let n = a.len();

    for i in 0..n {
        let diag = a[i][i].abs();
        let sum: f64 = (0..n)
            .filter(|&j| j != i)
            .map(|j| a[i][j].abs())
            .sum();

        print!("Строка {}: |a[{}][{}]| = {:.4}, сумма остальных = {:.4}", i + 1, i + 1, i + 1, diag, sum);
        if diag > sum {
            println!("  -> OK");
        } else {
            println!("  -> НЕ выполняется");
            ok = false;
        }
    }

    if ok {
        println!("\nУсловие сходимости выполняется.");
    } else {
        println!("\nУсловие сходимости НЕ выполняется.");
    }
}

fn seidel(a: &[Vec<f64>], b: &[f64], eps: f64, max_iter: usize) -> Vec<f64> {
    let n = a.len();
    let mut x = vec![0.0; n];
    let mut old = vec![0.0; n];

    println!("-------------------------------------------------------------");
    println!(" N        x1        x2        x3        x4        eps");
    println!("-------------------------------------------------------------");

    for iter in 1..=max_iter {
        old.copy_from_slice(&x);
        for i in 0..n {
            let mut sum = b[i];
            for j in 0..n {
                if j < i {
                    sum -= a[i][j] * x[j];
                } else if j > i {
                    sum -= a[i][j] * old[j];
                }
            }
            x[i] = sum / a[i][i];
        }

        let mut err = 0.0;
        for i in 0..n {
            err = err.max((x[i] - old[i]).abs());
        }

        println!("{:2} {:10.4} {:10.4} {:10.4} {:10.4} {:12.4}", iter, x[0], x[1], x[2], x[3], err);

        if err < eps {
            break;
        }
    }
    x
}

fn main() {
    let m_val = 0.89;
    let n_val = 0.12;
    let p_val = -1.15;

    let a = vec![
        vec![m_val, -0.04, 0.21, -1.16],
        vec![0.25, -1.23, n_val, -0.09],
        vec![-0.21, n_val, 0.8, -0.13],
        vec![0.15, -1.31, 0.06, p_val],
    ];
    let b = vec![-1.24, p_val, 2.56, m_val];

    println!("\nМетод Гаусса:");
    let xg = gauss(&a, &b);
    println!("\nРешение (Гаусс): x1={:.4}, x2={:.4}, x3={:.4}, x4={:.4}\n", xg[0], xg[1], xg[2], xg[3]);

    check_diagonal_dominance(&a);

    println!("\nМетод Зейделя:");
    let xs = seidel(&a, &b, 1e-3, 100);
    println!("\nРешение (Зейдель): x1={:.4}, x2={:.4}, x3={:.4}, x4={:.4}", xs[0], xs[1], xs[2], xs[3]);
}